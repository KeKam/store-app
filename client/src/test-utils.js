import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
