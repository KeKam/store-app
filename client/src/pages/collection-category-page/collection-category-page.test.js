import React from 'react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import CollectionCategoryPage from './collection-category-page';

const history = createMemoryHistory();

describe('CollectionCategoryPage component', () => {
  let mockMatch;

  beforeEach(() => {
    mockMatch = {
      isExact: true,
      params: {
        categoryId: 'spring',
      },
      path: '/collection/:categoryId',
      url: '/collection/spring',
    };
  });

  it('should render CollectionCategoryPage component', () => {
    const mockState = {
      categories: {
        spring: {
          id: 1,
          items: [
            {
              id: 1,
              imageUrl: 'www.testImage.com',
              name: 'Tulips',
              price: 20,
            },
          ],
          routeName: 'spring',
          title: 'Spring',
        },
      },
    };

    const { asFragment } = render(
      <Router history={history}>
        <CollectionCategoryPage match={mockMatch} />
      </Router>,
      { initialState: { collection: mockState } }
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('heading', { name: 'SPRING' })).toBeInTheDocument();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Tulips');
    expect(screen.getByTestId('price')).toHaveTextContent('20 â‚¬');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
    expect(screen.queryByTestId('spinner')).not.toBeInTheDocument();
  });

  it('should render Spinner component', () => {
    render(
      <Router history={history}>
        <CollectionCategoryPage match={mockMatch} />
      </Router>
    );

    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });
});
