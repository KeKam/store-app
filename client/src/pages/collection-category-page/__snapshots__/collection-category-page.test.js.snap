// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionCategoryPage component should render CollectionCategoryPage component 1`] = `
<DocumentFragment>
  <div
    class="sc-gZMcBi esSNAH"
  >
    <h2
      class="sc-gqjmRU kUdzW"
    >
      SPRING
    </h2>
    <div
      class="sc-VigVT jzxwRK"
    >
      <div
        class="sc-bwzfXH srITS"
      >
        <div
          class="sc-htpNat dTlvyq"
          data-testid="image"
        />
        <div
          class="sc-bxivhb oDKzY"
        >
          <span
            class="sc-EHOje ckFQmK"
            data-testid="title"
          >
            Tulips
          </span>
          <span
            class="sc-bZQynM bjXVPL"
            data-testid="price"
          >
            20 â‚¬
          </span>
        </div>
        <a
          class="sc-gzVnrw jnQMmI"
          data-testid="link"
          href="/login"
        >
          <h2
            class="sc-htoDjs bRjjzo"
          >
            Sign in required
          </h2>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
